rig-core/src/providers/mira.rs

```diff
@@ -0,0 +1,793 @@
+    Structured {
+        id: String,
+        object: String,
+        created: u64,
+        model: String,
+        choices: Vec<ChatChoice>,
+        #[serde(skip_serializing_if = "Option::is_none")]
+        usage: Option<Usage>,
+    },
+    Simple(String),
+}
+
+#[derive(Debug, Deserialize)]
+pub struct ChatChoice {
+    #[serde(deserialize_with = "deserialize_message")]
+    pub message: message::Message,
+    #[serde(default)]
+    pub finish_reason: Option<String>,
+    #[serde(default)]
+    pub index: Option<usize>,
+}
+
+fn deserialize_message<'de, D>(deserializer: D) -> Result<message::Message, D::Error>
+where
+    D: serde::Deserializer<'de>,
+{
+    let raw = RawMessage::deserialize(deserializer)?;
+    message::Message::try_from(raw).map_err(serde::de::Error::custom)
+}
+
+#[derive(Debug, Deserialize)]
+struct ModelsResponse {
+    data: Vec<ModelInfo>,
+}
+
+#[derive(Debug, Deserialize)]
+struct ModelInfo {
+    id: String,
+}
+
+#[derive(Clone)]
+/// Client for interacting with the Mira API
+pub struct Client {
+    base_url: String,
+    client: reqwest::Client,
+    headers: HeaderMap,
+}
+
+impl Client {
+    /// Create a new Mira client with the given API key
+    pub fn new(api_key: &str) -> Result<Self, MiraError> {
+        let mut headers = HeaderMap::new();
+        headers.insert(CONTENT_TYPE, HeaderValue::from_static("application/json"));
+        headers.insert(
+            AUTHORIZATION,
+            HeaderValue::from_str(&format!("Bearer {}", api_key))
+                .map_err(|_| MiraError::InvalidApiKey)?,
+        );
+        headers.insert(
+            reqwest::header::ACCEPT,
+            HeaderValue::from_static("application/json"),
+        );
+        headers.insert(
+            reqwest::header::USER_AGENT,
+            HeaderValue::from_static("rig-client/1.0"),
+        );
+
+        Ok(Self {
+            base_url: MIRA_API_BASE_URL.to_string(),
+            client: reqwest::Client::builder()
+                .connect_timeout(std::time::Duration::from_secs(30))
+                .build()
+                .expect("Failed to build HTTP client"),
+            headers,
+        })
+    }
+
+    /// Create a new Mira client from the `MIRA_API_KEY` environment variable.
+    /// Panics if the environment variable is not set.
+    pub fn from_env() -> Result<Self, MiraError> {
+        let api_key = std::env::var("MIRA_API_KEY").expect("MIRA_API_KEY not set");
+        Self::new(&api_key)
+    }
+
+    /// Create a new Mira client with a custom base URL and API key
+    pub fn new_with_base_url(
+        api_key: &str,
+        base_url: impl Into<String>,
+    ) -> Result<Self, MiraError> {
+        let mut client = Self::new(api_key)?;
+        client.base_url = base_url.into();
+        Ok(client)
+    }
+
+    /// Generate a chat completion
+    pub async fn generate(
+        &self,
+        model: &str,
+        request: CompletionRequest,
+    ) -> Result<CompletionResponse, MiraError> {
```

Comment:
It's unclear why this exists. The main way to generate responses from `CompletionRequest`s is through the `.completion` api defined by the `CompletionModel` trait. This code is nearly duplicated below.